src/librustc/middle/liveness.rs

```diff
@@ -1479,7 +1479,10 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
                 self.ir.tcx.region_maps.call_site_extent(id, body.id),
                 &self.fn_ret(id));
 
-        if self.live_on_entry(entry_ln, self.s.no_ret_var).is_some() {
+        if fn_ret.is_never() && self.live_on_entry(entry_ln, self.s.clean_exit_var).is_some() {
```

Comment:
These should not be combined, i.e. the `else` below should be to `fn_ret.is_never()` alone, to match the old semantics.